File: app/src/main/java/es/etologic/mahjongscoring2/app/old_games/GameItemDiffUtilCallback.java
package es.etologic.mahjongscoring2.app.old_games;

import android.support.v7.util.DiffUtil;

import java.util.List;

import es.etologic.mahjongscoring2.app.utils.DateUtils;
import es.etologic.mahjongscoring2.domain.entities.BestHand;
import es.etologic.mahjongscoring2.domain.entities.Game;

class GameItemDiffUtilCallback extends DiffUtil.Callback {

    private static final int NUM_GAME_PLAYERS = 4;
    private List<Game> oldList;
    private List<Game> newList;

    GameItemDiffUtilCallback(List<Game> newList, List<Game> oldList) {
        this.newList = newList;
        this.oldList = oldList;
    }


    @Override
        public int getOldListSize() {
            return oldList.size();
        }

        @Override
        public int getNewListSize() {
            return newList.size();
        }

        @Override
        public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
            return oldList.get(oldItemPosition).getGameId() ==
                    newList.get(newItemPosition).getGameId();
        }

        @Override
        public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
            Game newGame = newList.get(newItemPosition);
            Game oldGame = oldList.get(oldItemPosition);
            return  oldGame.getGameId() == newGame.getGameId() &&
                    oldGame.getNameP1().equals(newGame.getNameP1()) &&
                    oldGame.getNameP2().equals(newGame.getNameP2()) &&
                    oldGame.getNameP3().equals(newGame.getNameP3()) &&
                    oldGame.getNameP4().equals(newGame.getNameP4()) &&
                    DateUtils.areEqual(oldGame.getStartDate(), newGame.getStartDate()) &&
                    DateUtils.areEqual(oldGame.getEndDate(), newGame.getEndDate()) &&
                    arePlayersTotalsPointsEquals(oldGame.getPlayersTotalPoints(),
                            newGame.getPlayersTotalPoints()) &&
                    areBestHandsEqual(oldGame.getBestHand(), newGame.getBestHand());
        }

    private boolean arePlayersTotalsPointsEquals(String[] oldPlayersTotalPoints,
                                                 String[] newPlayersTotalPoints) {
        for(int i = 0; i < NUM_GAME_PLAYERS; i++) {
            if(!oldPlayersTotalPoints[i].equals(newPlayersTotalPoints[i])) {
                return false;
            }
        }
        return true;
    }

    private boolean areBestHandsEqual(BestHand oldBestHand, BestHand newBestHand) {
        return oldBestHand.getPlayerName().equals(newBestHand.getPlayerName()) &&
                oldBestHand.getHandValue() == newBestHand.getHandValue();
    }
}

File: app/src/main/java/com/etologic/mahjongscoring2/app/model/GameItemDiffUtilCallback.kt
/*
    Copyright (C) 2020  Ernesto Vega de la Iglesia Soria
 
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package com.etologic.mahjongscoring2.app.model

import androidx.recyclerview.widget.DiffUtil
import com.etologic.mahjongscoring2.app.utils.DateTimeUtils
import com.etologic.mahjongscoring2.business.model.dtos.BestHand
import com.etologic.mahjongscoring2.business.model.entities.Round
import com.etologic.mahjongscoring2.business.model.entities.Table
import com.etologic.mahjongscoring2.business.model.entities.Table.Companion.NUM_MCR_PLAYERS
import java.util.Locale

class GameItemDiffUtilCallback(private val newList: List<Table>, private val oldList: List<Table>) : DiffUtil.Callback() {

    override fun getOldListSize(): Int {
        return oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldList[oldItemPosition].game.gameId == newList[newItemPosition].game.gameId
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val newGame = newList[newItemPosition]
        val oldGame = oldList[oldItemPosition]
        return (oldGame.game.gameId == newGame.game.gameId &&
                oldGame.game.nameP1 == newGame.game.nameP1 &&
                oldGame.game.nameP2 == newGame.game.nameP2 &&
                oldGame.game.nameP3 == newGame.game.nameP3 &&
                oldGame.game.nameP4 == newGame.game.nameP4 &&
                DateTimeUtils.areEqual(oldGame.game.startDate, newGame.game.startDate) &&
                arePlayersTotalsPointsEquals(
                    oldGame.getPlayersTotalPointsByCurrentSeat().map {
                        String.format(Locale.getDefault(), "%d", it)
                    },
                    newGame.getPlayersTotalPointsByCurrentSeat().map {
                        String.format(Locale.getDefault(), "%d", it)
                    }
                ) &&
                areBestHandsEqual(oldGame.getBestHand(), newGame.getBestHand())
                && Round.areEqual(oldGame.rounds, newGame.rounds))
    }

    private fun arePlayersTotalsPointsEquals(
        oldPlayersTotalPoints: List<String>,
        newPlayersTotalPoints: List<String>
    ): Boolean {
        for (i in 0 until NUM_MCR_PLAYERS) {
            if (oldPlayersTotalPoints[i] != newPlayersTotalPoints[i]) {
                return false
            }
        }
        return true
    }

    private fun areBestHandsEqual(oldBestHand: BestHand, newBestHand: BestHand): Boolean {
        return oldBestHand.playerName == newBestHand.playerName && oldBestHand.handValue == newBestHand.handValue
    }
}
